File: src/types/usageLimit.ts
// API Key 使用限额（Budget Limit）类型定义
//
// 与后端 `services/usage_limit.rs` 的 serde 输出一一对应（camelCase）。

/** 限额模式：同一时间只有一种 */
export type UsageLimitType = "money" | "token";

/**
 * 金额模式币种（V1.0.2 起支持 5 种主流货币）。
 * 数据库内部成本统一为 USD，其他币种通过本地可配置汇率换算（人工调节，不联网）。
 */
export type UsageLimitCurrency = "USD" | "CNY" | "EUR" | "JPY" | "GBP";

/** 各币种展示符号（与后端 LimitCurrency::symbol 一致；JPY 加国别前缀消歧） */
export const USAGE_LIMIT_CURRENCY_SYMBOLS: Record<UsageLimitCurrency, string> =
  {
    USD: "$",
    CNY: "¥",
    EUR: "€",
    JPY: "JP¥",
    GBP: "£",
  };

/**
 * 统计窗口重置周期（V1.0.1）。
 * never = 仅手动重置；其余在本地时区周期边界（整点/零点/周一零点/每月 1 日零点）
 * 自动重置——后端为懒滚动实现，应用未运行期间跨过的边界在下一次启动时照常生效。
 */
export type UsageLimitResetPeriod =
  | "never"
  | "hourly"
  | "daily"
  | "weekly"
  | "monthly";

/** enforcement 能力：Proxy 是否真的在该 Provider 的请求路径上 */
export type UsageLimitEnforcement =
  | "active" // Proxy 已接管该应用，限额可强制执行
  | "proxy_disabled" // 该应用 Proxy 接管未开启，流量不经过 CC Switch
  | "unsupported_credential"; // OAuth 类 Provider，无静态 Key

/** 卡片图标 / Dialog 共用的状态聚合 */
export interface UsageLimitStatus {
  providerId: string;
  appType: string;
  enabled: boolean;
  limitType?: UsageLimitType | null;
  currency?: UsageLimitCurrency | null;
  limitAmount?: string | null;
  /** 预算统计窗口起点（unix 秒；周期重置下为对齐周期边界后的有效起点） */
  usageStartAt?: number | null;
  /** 重置周期（V1.0.1），缺省视为 never */
  resetPeriod?: UsageLimitResetPeriod | null;
  /** 下一次周期重置时间（unix 秒）；never 时为 null */
  nextResetAt?: number | null;
  /** 窗口内已用金额（USD 原值，十进制字符串） */
  usedMoneyUsd: string;
  /** 限额币种下的已用金额（CNY 限额时已按汇率换算） */
  usedMoneyInCurrency?: string | null;
  /** 窗口内已用 token（与 Usage Dashboard 一致的 normalized total） */
  usedTokens: number;
  /** 使用百分比字符串（十进制，不 clamp，可超过 100） */
  percentUsed?: string | null;
  /** "off" | "active" | "warning" | "exhausted" */
  state: UsageLimitState;
  /** 窗口内「有 token 用量但当前无定价」的请求数（金额模式可靠性提示） */
  unpricedRequestCount: number;
  /** 上述请求涉及的模型列表（引导配置 custom pricing） */
  unpricedModels: string[];
  enforcement: UsageLimitEnforcement;
  /** 当前 credential 遮蔽展示（如 sk-****ABCD），无静态 key 时为 null */
  maskedCredential?: string | null;
}

export type UsageLimitState = "off" | "active" | "warning" | "exhausted";

/** 保存/开关提交载荷 */
export interface UsageLimitConfig {
  enabled: boolean;
  limitType: UsageLimitType;
  /** token 模式下为 null */
  currency?: UsageLimitCurrency | null;
  /** 金额（十进制字符串）或 token 数（正整数字符串） */
  limitAmount?: string | null;
  /** 重置周期（V1.0.1），始终显式提交 */
  resetPeriod: UsageLimitResetPeriod;
}

File: src-tauri/src/database/dao/usage_limit.rs
//! API Key 使用限额 DAO
//!
//! `api_key_limits` 表的持久化访问，以及从 usage SSOT（`proxy_request_logs`）
//! 聚合指定 credential 在预算窗口内的用量。
//!
//! 设计要点：
//! - 本表只保存限额配置与统计窗口起点（`usage_start_at`），不冗余存储
//!   累计用量（避免与 `proxy_request_logs` 形成双 SSOT）。
//! - 金额比较走两级通道：先 REAL 快速判（远离阈值时 O(1) 放行），贴近
//!   阈值时取原始十进制字符串用 `rust_decimal` 精确求和复核，规避浮点误差。

use crate::database::{lock_conn, Database};
use crate::error::AppError;
use rusqlite::{params, OptionalExtension};

/// 一条 API Key 限额配置
#[derive(Debug, Clone, PartialEq)]
pub struct ApiKeyLimitRow {
    pub provider_id: String,
    pub app_type: String,
    /// API Key 的不可逆 SHA-256 指纹（十六进制），绝不存明文 key
    pub credential_fingerprint: String,
    pub enabled: bool,
    /// "money" 或 "token"
    pub limit_type: String,
    /// "USD" / "CNY" / "EUR" / "JPY" / "GBP"；token 模式下为 None
    pub currency: Option<String>,
    /// 十进制字符串（金额）或纯整数字符串（token 数）
    pub limit_amount: String,
    /// 预算统计窗口起点（unix 秒）
    pub usage_start_at: i64,
    /// 重置周期："never" | "hourly" | "daily" | "weekly" | "monthly" | "custom"
    pub reset_period: String,
    /// 自定义窗口长度（仅 reset_period = 'custom' 时有意义）
    pub window_length: Option<i64>,
    /// 自定义窗口单位："hours" | "days"
    pub window_unit: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

/// 指定 credential 在窗口内的用量聚合（REAL 快速通道）
#[derive(Debug, Clone, Default)]
pub struct CredentialUsageFast {
    /// 窗口内 `total_cost_usd` 的 REAL 近似和（仅用于快速预判）
    pub total_cost_usd_real: f64,
    /// 归一化 fresh input tokens（与 Usage Dashboard 同一语义）
    pub fresh_input_tokens: i64,
    pub output_tokens: i64,
    pub cache_creation_tokens: i64,
    pub cache_read_tokens: i64,
}

impl CredentialUsageFast {
    /// Dashboard 语义的 normalized total tokens：
    /// fresh_input + output + cache_creation + cache_read
    pub fn normalized_total_tokens(&self) -> i64 {
        self.fresh_input_tokens
            .saturating_add(self.output_tokens)
            .saturating_add(self.cache_creation_tokens)
            .saturating_add(self.cache_read_tokens)
    }
}

impl Database {
    /// 读取某个 Provider 的限额配置（每个 Provider 至多一条）
    pub fn get_api_key_limit(
        &self,
        provider_id: &str,
        app_type: &str,
    ) -> Result<Option<ApiKeyLimitRow>, AppError> {
        let conn = lock_conn!(self.conn);
        conn.query_row(
            "SELECT provider_id, app_type, credential_fingerprint, enabled, limit_type,
                    currency, limit_amount, usage_start_at, reset_period, window_length,
                    window_unit, created_at, updated_at
             FROM api_key_limits WHERE provider_id = ?1 AND app_type = ?2",
            params![provider_id, app_type],
            |row| {
                Ok(ApiKeyLimitRow {
                    provider_id: row.get(0)?,
                    app_type: row.get(1)?,
                    credential_fingerprint: row.get(2)?,
                    enabled: row.get::<_, i64>(3)? != 0,
                    limit_type: row.get(4)?,
                    currency: row.get(5)?,
                    limit_amount: row.get(6)?,
                    usage_start_at: row.get(7)?,
                    reset_period: row.get(8)?,
                    window_length: row.get(9)?,
                    window_unit: row.get(10)?,
                    created_at: row.get(11)?,
                    updated_at: row.get(12)?,
                })
            },
        )
        .optional()
        .map_err(|e| AppError::Database(format!("读取限额配置失败: {e}")))
    }

    /// 保存限额配置（首次插入时由调用方决定 usage_start_at）
    pub fn upsert_api_key_limit(&self, row: &ApiKeyLimitRow) -> Result<(), AppError> {
        let conn = lock_conn!(self.conn);
        conn.execute(
            "INSERT INTO api_key_limits (
                provider_id, app_type, credential_fingerprint, enabled, limit_type,
                currency, limit_amount, usage_start_at, reset_period, window_length,
                window_unit, created_at, updated_at
             ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13)
             ON CONFLICT (provider_id, app_type) DO UPDATE SET
                credential_fingerprint = excluded.credential_fingerprint,
                enabled = excluded.enabled,
                limit_type = excluded.limit_type,
                currency = excluded.currency,
                limit_amount = excluded.limit_amount,
                usage_start_at = excluded.usage_start_at,
                reset_period = excluded.reset_period,
                window_length = excluded.window_length,
                window_unit = excluded.window_unit,
                updated_at = excluded.updated_at",
            params![
                row.provider_id,
                row.app_type,
                row.credential_fingerprint,
                row.enabled as i64,
                row.limit_type,
                row.currency,
                row.limit_amount,
                row.usage_start_at,
                row.reset_period,
                row.window_length,
                row.window_unit,
                row.created_at,
                row.updated_at,
            ],
        )
        .map_err(|e| AppError::Database(format!("保存限额配置失败: {e}")))?;
        Ok(())
    }

    /// 重置预算窗口：usage_start_at 推进到当前时间，不删除任何历史 Usage
    pub fn reset_api_key_limit_window(
        &self,
        provider_id: &str,
        app_type: &str,
        new_start: i64,
    ) -> Result<(), AppError> {
        let conn = lock_conn!(self.conn);
        let affected = conn
            .execute(
                "UPDATE api_key_limits SET usage_start_at = ?3, updated_at = ?3
                 WHERE provider_id = ?1 AND app_type = ?2",
                params![provider_id, app_type, new_start],
            )
            .map_err(|e| AppError::Database(format!("重置使用量失败: {e}")))?;
        if affected == 0 {
            return Err(AppError::InvalidInput(
                "限额配置不存在，无法重置使用量".to_string(),
            ));
        }
        Ok(())
    }

    /// Credential 轮换重绑：Provider 换了 API Key 时，把配置迁移到新指纹并
    /// 将统计窗口起点重置为当前时间——旧 Key 的历史用量不会计入新 Key。
    /// 返回 true 表示确实发生了重绑。
    pub fn rebind_api_key_limit_credential(
        &self,
        provider_id: &str,
        app_type: &str,
        fingerprint: &str,
        now: i64,
    ) -> Result<bool, AppError> {
        let conn = lock_conn!(self.conn);
        let affected = conn
            .execute(
                "UPDATE api_key_limits
                 SET credential_fingerprint = ?3, usage_start_at = ?4, updated_at = ?4
                 WHERE provider_id = ?1 AND app_type = ?2 AND credential_fingerprint <> ?3",
                params![provider_id, app_type, fingerprint, now],
            )
            .map_err(|e| AppError::Database(format!("重绑限额凭证失败: {e}")))?;
        Ok(affected > 0)
    }

    /// 用量快速聚合（REAL 金额近似 + 精确整数 token 和）。
    /// 只统计 `data_source = 'proxy'` 的请求：预算限额是本地代理观测到的流量的边界。
    pub fn sum_credential_usage_fast(
        &self,
        provider_id: &str,
        app_type: &str,
        fingerprint: &str,
        since: i64,
    ) -> Result<CredentialUsageFast, AppError> {
        let conn = lock_conn!(self.conn);
        let fresh_input = crate::services::sql_helpers::fresh_input_sql("");
        let sql = format!(
            "SELECT COALESCE(SUM(CAST(total_cost_usd AS REAL)), 0),
                    COALESCE(SUM({fresh_input}), 0),
                    COALESCE(SUM(output_tokens), 0),
                    COALESCE(SUM(cache_creation_tokens), 0),
                    COALESCE(SUM(cache_read_tokens), 0)
             FROM proxy_request_logs
             WHERE provider_id = ?1 AND app_type = ?2
               AND credential_fingerprint = ?3 AND created_at >= ?4
               AND data_source = 'proxy'"
        );
        conn.query_row(
            &sql,
            params![provider_id, app_type, fingerprint, since],
            |row| {
                Ok(CredentialUsageFast {
                    total_cost_usd_real: row.get(0)?,
                    fresh_input_tokens: row.get(1)?,
                    output_tokens: row.get(2)?,
                    cache_creation_tokens: row.get(3)?,
                    cache_read_tokens: row.get(4)?,
                })
            },
        )
        .map_err(|e| AppError::Database(format!("聚合凭证用量失败: {e}")))
    }

    /// 金额精确复核：取窗口内原始 `total_cost_usd` 十进制字符串，
    /// 由调用方用 `rust_decimal` 精确求和（贴近阈值时才调用）。
    pub fn fetch_credential_cost_strings(
        &self,
        provider_id: &str,
        app_type: &str,
        fingerprint: &str,
        since: i64,
    ) -> Result<Vec<String>, AppError> {
        let conn = lock_conn!(self.conn);
        let mut stmt = conn
            .prepare(
                "SELECT total_cost_usd FROM proxy_request_logs
                 WHERE provider_id = ?1 AND app_type = ?2
                   AND credential_fingerprint = ?3 AND created_at >= ?4
                   AND data_source = 'proxy'",
            )
            .map_err(|e| AppError::Database(e.to_string()))?;
        let rows = stmt
            .query_map(params![provider_id, app_type, fingerprint, since], |row| {
                row.get::<_, String>(0)
            })
            .map_err(|e| AppError::Database(e.to_string()))?;
        let mut costs = Vec::new();
        for row in rows {
            costs.push(row.map_err(|e| AppError::Database(e.to_string()))?);
        }
        Ok(costs)
    }

    /// 找出窗口内「有 token 用量但成本为 0」的请求涉及的计价模型，
    /// 供上层逐个检查是否真的缺少定价（免费模型价格 0 不算缺价）。
    pub fn list_zero_cost_pricing_models(
        &self,
        provider_id: &str,
        app_type: &str,
        fingerprint: &str,
        since: i64,
    ) -> Result<Vec<(String, i64)>, AppError> {
        let conn = lock_conn!(self.conn);
        let mut stmt = conn
            .prepare(
                "SELECT pricing_model, COUNT(*) FROM proxy_request_logs
                 WHERE provider_id = ?1 AND app_type = ?2
                   AND credential_fingerprint = ?3 AND created_at >= ?4
                   AND data_source = 'proxy'
                   AND (input_tokens > 0 OR output_tokens > 0
                        OR cache_read_tokens > 0 OR cache_creation_tokens > 0)
                   AND total_cost_usd = '0'
                   AND IFNULL(pricing_model, '') <> ''
                 GROUP BY pricing_model",
            )
            .map_err(|e| AppError::Database(e.to_string()))?;
        let rows = stmt
            .query_map(params![provider_id, app_type, fingerprint, since], |row| {
                Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?))
            })
            .map_err(|e| AppError::Database(e.to_string()))?;
        let mut models = Vec::new();
        for row in rows {
            models.push(row.map_err(|e| AppError::Database(e.to_string()))?);
        }
        Ok(models)
    }
}
